snuba/utils/streams/batching.py

```diff
@@ -161,6 +161,15 @@ def signal_shutdown(self) -> None:
     def _handle_message(self, msg: Message[TPayload]) -> None:
         start = time.time()
 
+        self.__metrics.timing(
+            "receive_latency",
+            (start - msg.timestamp.timestamp()) * 1000,
+            tags={
+                "topic": msg.partition.topic.name,
+                "partition": str(msg.partition.index),
+            },
+        )
```

Comment:
For each message, I think it would be useful to track:

- receive latency (how long it took to receive the message, may or may not include codec.decode over the long term, not sure; codec.decode should be tracked separately)
- processing latency (how long it took to "process" the message, where the specific definition of "process" is up to the implementation, for subscriptions this is how long it took to run the quer; for conventional/existing processors this would include building the row to write, it could also include the socket write on a per-row basis if we wanted)
- committed latency (how long it took before this message was committed to Kafka)

The way the consumer is structured right now, processing latency would be misleading for the subscriptions consumer as soon as we actually start using the threaded executor for asynchronous queries, and committed latency is generally a challenge with this structure.

I'm considering taking a stab at making the batching consumer support asynchronous processing by default (processors would return futures, whether processing happens synchronously or asynchronously.) I don't think it makes sense to try to implement those other metrics until that point.

Reply:
> I'm considering taking a stab at making the batching consumer support asynchronous processing by default (processors would return futures, whether processing happens synchronously or asynchronously.)

This would also hypothetically enable a much simpler transition to any sort of multiprocess consumer architecture if we still want to do that down the line.